File: Easy/Python/104. Maximum Depth of Binary Tree.py
#Given a binary tree, find its maximum depth.
#The maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.
#Note: A leaf is a node with no children.

#Runtime: (48ms) Faster than  87.69% of Python submissions; MEM: (15.3mb) 

class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        return 1 +max(self.maxDepth(root.left), self.maxDepth(root.right)) if root else 0
      

File: Easy/C++/104. Maximum Depth of Binary Tree.cpp
/*
Given a binary tree, find its maximum depth.
The maximum depth is the number of nodes along the longest path from the root node down to the 
farthest leaf node.

Runtime: beats 98.99 % of cpp submissions. 
MEM: Your memory usage beats 96.47 % of cpp submissions.
*/

class Solution {
public:
	int maxDepth(TreeNode* root) {
		//do in level traversal        
		//initialize
		queue<TreeNode*> curr_level;
		queue<TreeNode*> next_level;

		if (!root) return 0;
		else next_level.push(root);
		int depth = 0;

		while (!next_level.empty()) {
			swap(curr_level, next_level);
			cout << "1";
			++depth;
			while (!curr_level.empty()) {
				TreeNode*& node = curr_level.front();
				if (node->left) next_level.push(node->left);
				if (node->right) next_level.push(node->right);
				curr_level.pop();
			}
		}
		return depth;
	}
};